### PR TITLE
Reduce cyclocomp of `lint()` to <= 15

### LINKED ISSUE #1186: Cyclomatic complexity lints in lintr
After fixing all implicit integer lints, we can almost use `.lintr_new`, except for cyclomatic complexity:

 - [ ] `lint`: 40 -> 29
 - [x] `lint_parse_error`: 24
 - [x] `commas_linter`: 23
 - [x] `object_usage_linter`: 21
 - [x] `sprintf_linter`: 19
 - [x] `normalize_exclusions`: 18
 - [x] `fix_eq_assigns`: 18
 - [x] `available_linters`: 18
 - [x] `namespace_linter`: 18
 - [x] `closed_curly_linter`: 17
 - [x] `split_path`: 17
 - [x] `fix_tab_indentations`: 16
 - [x] `lint_dir`: 16
 - [x] `open_curly_linter`: 16

### PR DESCRIPTION
**Pull request by AshesITR**
Closes #1186

### CODE REVIEW

File: R/lint.R [hunk 1]
```diff
@@ -114,14 +117,7 @@ lint <- function(filename, linters = NULL, ..., cache = FALSE, parse_settings =
...
 
-  if (inherits(source_expressions$error, "lint")) {
-    lints[[itr <- itr + 1L]] <- source_expressions$error
-
-    if (isTRUE(cache)) {
-      cache_lint(lint_cache, list(filename = filename, content = ""), "error", source_expressions$error)
-    }
-  }
-
+  lints <- maybe_append_error_lint(lints, source_expressions$error, lint_cache, filename)
```
Comment (MichaelChirico) [inline R/lint.R:120]: could you double-check for any performance implications here? this'll copy the list whenever an edit is needed

Comment (AshesITR) [reply R/lint.R:120]: Will do once refactoring is complete.

File: R/lint.R [hunk 2]
```diff
@@ -114,14 +100,7 @@ lint <- function(filename, linters = NULL, ..., cache = FALSE, parse_settings =
     }
```
Comment (MichaelChirico) [inline R/lint.R:100]: do this as `mapply()`?

File: R/lint.R [hunk 3]
```diff
@@ -42,38 +42,24 @@
...
-    lines <- strsplit(text, "\n", fixed = TRUE)[[1L]]
-  }
 
-  no_filename <- missing(filename) || is.null(filename)
+  needs_tempfile <- missing(filename) || rex::re_matches(filename, rex::rex(newline))
+  inline_data <- !is.null(text) || needs_tempfile
+  lines <- get_lines(filename, text)
 
-  if (inline_data && no_filename) {
+  if (needs_tempfile) {
```
Comment (MichaelChirico) [inline R/lint.R:58]: do this in a helper? helper returns temp filename so `on.exit(unlink())` happens fluidly

Comment (AshesITR) [reply R/lint.R:58]: How do you mean it?

File: R/lint.R (hunk 2 above)
Comment (AshesITR) [reply R/lint.R:92]: The list grows to `length(source_expressions$expressions) * length(linters)`. Could be a case for `outer()`...
Not related to cyclocomp though - for loops are linear.

**Comment by AshesITR** [comment]
Okay, we're there. PTAL.
If it looks good, I'll do a quick performance comparison - the test suite runtime hasn't changed much, but I'll check on `tools/QC.R` and `default_linter_testcode.R` tomorrow if you think the code looks good.

File: R/lint.R (hunk 2 above)
Comment (MichaelChirico) [reply R/lint.R:92]: i don't think that's true:

```
cyclocomp::cyclocomp(function() for (i in 1:10) i)
# [1] 3
cyclocomp::cyclocomp(function() for (i in 1:10) for (j in 1:10) i*j)
# [1] 6
cyclocomp::cyclocomp(function() for (i in 1:10) for (j in 1:10) for (k in 1:10) i*j*k)
# [1] 10
cyclocomp::cyclocomp(function() for (i in 1:10) for (j in 1:10) for (k in 1:10) for (l in 1:10) i*j*k*l)
# [1] 15
```

looks fibonacci=quadratic actually

File: R/lint.R (hunk 3 above)
Comment (MichaelChirico) [reply R/lint.R:58]:
```r
# helper
write_text_to_tempfile <- function(lines) {
  filename <- tempfile()
  con <- file(filename, open = "w", encoding = settings$encoding)
  on.exit(close(con))
  writeLines(text = lines, con = con, sep = "\n")
  filename
}

# usage
if (needs_tempfile) {
  filename <- write_text_to_tempfile(lines)
  on.exit(unlink(lines), add = TRUE
}
```

i guess that doesn't help cyclocomp unless we consume the `if()` into the helper :|

**Comment by MichaelChirico** [comment]
Reading more carefully, I think it's fine. The copy is only needed in case of an error; I'm OK with prioritizing readability over pure efficiency in error regions.

**Review by MichaelChirico — DISMISSED** [review]

**Comment by MichaelChirico** [comment]
Will leave you to merge

File: R/lint.R (hunk 2 above)
Comment (AshesITR) [reply R/lint.R:92]: Oh I didn't know that. Thanks.
Since we're already below 15, do you still want to try an `outer()` based approach?

Comment (MichaelChirico) [reply R/lint.R:92]: i think it'll look nicer / more "canonically" R, but not a priority

**Review by MichaelChirico — APPROVED** [review]

**Comment by AshesITR** [comment]
Performance is unaffected for `default_linter_testcode.R`:
```
> bench::mark(lintr::lint("tests/testthat/default_linter_testcode.R", linters = lintr::linters_with_tags(NULL), parse_settings = FALSE), min_iterations = 5)

# main
  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result  memory time           gc      
  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>  <list> <list>         <list>  
1 730ms    758ms      1.21        NA     13.0     5    54      4.15s <lints> <NULL> <bench_tm [5]> <tibble>

# this PR
  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result  memory time           gc      
  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>  <list> <list>         <list>  
1 726ms    749ms      1.23        NA     14.5     5    59      4.08s <lints> <NULL> <bench_tm [5]> **<tibble**
```

In `tools/QC.R` I found a bug in `sprintf_linter()` for multi-line sprintf() calls with comments. #1249 
```
> bench::mark(
+   lintr::lint(test_file, linters = lintr::linters_with_tags(NULL, sprintf_linter = NULL), parse_settings = FALSE),
+   min_iterations = 5
+ )
# main
  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result  memory time           gc              
  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>  <list> <list>         <list>          
1 36.2s    36.8s    0.0271        NA     6.83     5  1259      3.07m <lints> <NULL> <bench_tm [5]> <tibble [5 × 3]>

# this PR
  min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result  memory time           gc              
  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>  <list> <list>         <list>          
1 37.4s    37.6s    0.0264        NA     7.23     5  1367      3.15m <lints> <NULL> <bench_tm [5]> <tibble [5 × 3]>
```

LGTM, so merging.

File: R/lint.R (hunk 3 above)
Comment (AshesITR) [reply R/lint.R:58]: Yeah, unfortunately it doesn't.
I mean we could be mean and do an `on.exit()` on `parent.frame()`, but I think that's a code smell not worth pursuing.

File: R/lint.R [hunk 4]
```diff
@@ -42,38 +42,24 @@
...
-  } else {
-    inline_data <- TRUE
-    if (length(text) > 1L) {
-      text <- paste(text, collapse = "\n")
-    }
-    lines <- strsplit(text, "\n", fixed = TRUE)[[1L]]
-  }
 
-  no_filename <- missing(filename) || is.null(filename)
+  needs_tempfile <- missing(filename) || rex::re_matches(filename, rex::rex(newline))
```
Comment (MichaelChirico) [inline R/lint.R:54]: @AshesITR I've been starting at this a while for #2135... why are `needs_tempfile` and `inline_data` different?

Comparing:

```r
needs_tempfile <=>                   missing(filename) || rex::re_matches(filename, rex::rex(newline))
inline_data    <=> !is.null(text) || missing(filename) || rex::re_matches(filename, rex::rex(newline))
```

But don't we need `text=` present when `missing(filename)`?

Comment (AshesITR) [reply R/lint.R:54]: IIRC there was something with a filename being present and content passed as well for linter integrations that lint an unsaved buffer version of a file. But I'm not sure.

Comment (MichaelChirico) [reply R/lint.R:54]: Hmm, in that case, I will table #2135 until after release. Once we release, we can try disabling this/combining the conditions and see if it affects anything.

**Pull request merged**